test/e2e/app/app.go

```diff
@@ -315,21 +323,25 @@ func (app *Application) CheckTx(_ context.Context, req *abci.CheckTxRequest) (*a
 		time.Sleep(app.cfg.CheckTxDelay)
 	}
 
-	// Assign a lane to the transaction deterministically.
-	var lane uint32
-	txHash := tmhash.Sum(req.Tx)
-	switch {
-	case txHash[0] == 0 && txHash[1] == 0 && txHash[2] == 0 && txHash[3] == 0:
-		lane = app.lanes["foo"]
-	case txHash[0] == 0:
-		lane = app.lanes["bar"]
-	default:
-		lane = app.lanes[defaultLane]
-	}
+	lane := extractLane(value)
 
 	return &abci.CheckTxResponse{Code: kvstore.CodeTypeOK, GasWanted: 1, Lane: lane}, nil
 }
 
+// extractLane returns the lane field if value is a Payload, otherwise returns 0.
+func extractLane(value string) uint32 {
+	valueBytes, err := hex.DecodeString(value)
+	if err != nil {
+		return 0
```

Comment:
Can we panic here? This should never happen, and could hide a problem hard to troubleshoot

Reply:
Done.